akka-javasdk-testkit/src/main/java/akka/javasdk/testkit/TelemetryReader.java

```diff
@@ -0,0 +1,183 @@
+   * tools to accomplish their tasks.
+   *
+   * @param debugId The debug identifier used to trace the operation
+   * @return A list of tool names in invocation order, or an empty list if no tools were found
+   */
+  @ApiMayChange
+  public List<String> getAgentTools(String debugId) {
+    List<SpanData> spanDatas = spansFor(debugId);
+
+    AttributeKey<String> agentToolAttributeKey = AttributeKey.stringKey("gen_ai.tool.name");
+
+    return collectAttributeValues(spanDatas, agentToolAttributeKey);
+  }
+
+  /**
+   * Retrieves the sequence of workflow steps executed for an operation traced by a debug ID.
+   *
+   * <p>Returns a list of step names in the order they were executed, based on telemetry data
+   * collected during workflow execution. This method uses the debug ID to find all spans associated
+   * with the traced operation, making it useful when you don't have direct access to the workflow
+   * class or workflow ID.
+   *
+   * @param debugId The debug identifier used to trace the workflow execution
+   * @return A list of step names in execution order, or an empty list if no steps were found
+   */
+  @ApiMayChange
+  public List<String> getWorkflowSteps(String debugId) {
+    List<SpanData> spanDatas = spansFor(debugId);
+
+    return workflowStepsFrom(spanDatas);
+  }
+
+  private List<String> collectAttributeValues(
+      List<SpanData> spanDatas, AttributeKey<String> agentToolAttributeKey) {
+    List<SpanData> stepSpanDatas = collectByAttribute(spanDatas, agentToolAttributeKey);
+
+    return stepSpanDatas.stream()
+        .sorted(Comparator.comparing(SpanData::getStartEpochNanos))
+        .map(spanData -> spanData.getAttributes().get(agentToolAttributeKey))
+        .toList();
+  }
+
+  private List<String> workflowStepsFrom(List<SpanData> spanDatas) {
+    AttributeKey<String> workflowStepNameAttribute =
+        AttributeKey.stringKey("akka.workflow.step.name");
+
+    return collectAttributeValues(spanDatas, workflowStepNameAttribute);
+  }
+
+  private String getComponentId(Class<?> componentClass) {
+    Component componentAnnotation = componentClass.getAnnotation(Component.class);
+    if (componentAnnotation != null) {
+      return componentAnnotation.id();
+    }
+
+    // Fallback to the old ComponentId annotation for backward compatibility
+    return componentClass.getAnnotation(ComponentId.class).value();
+  }
+
+  private List<SpanData> collectByAttribute(
+      List<SpanData> spanDatas, AttributeKey<String> attributeKey) {
+    return spanDatas.stream()
+        .filter(spanData -> spanData.getAttributes().get(attributeKey) != null)
+        .toList();
+  }
+
+  private List<SpanData> spansFor(
+      String componentId, AttributeKey<String> attributeKey, String attributeValue) {
+
+    return inMemorySpanExporter.getFinishedSpanItems().stream()
+        .filter(
+            spanData -> {
+              return ofNullable(
```

Comment:
The statically imported `ofNullable` doesn't read as well as explicit `Optional.ofNullable`.